### PR TITLE
Add cmake support for nvc++ as CUDA compiler

### PR DESCRIPTION
**Pull request by dalg24**
cc @scottbiersdorff

### CODE REVIEW

**Review by crtrott — APPROVED** [review]

File: cmake/kokkos_test_cxx_std.cmake [hunk 1]
```diff
@@ -128,8 +128,9 @@ IF(KOKKOS_ENABLE_CUDA)
     ELSEIF(CMAKE_CXX_EXTENSIONS)
       MESSAGE(FATAL_ERROR "Compiling CUDA code with clang doesn't support C++ extensions.  Set -DCMAKE_CXX_EXTENSIONS=OFF")
     ENDIF()
+  ELSEIF(KOKKOS_CXX_COMPILER_ID STREQUAL NVHPC)
   ELSEIF(NOT KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA)
```
Comment (masterleinad) [inline cmake/kokkos_test_cxx_std.cmake:132]:
```suggestion
  ELSEIF(NOT KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA AND NOT KOKKOS_CXX_COMPILER_ID STREQUAL NVHPC)
```

File: core/unit_test/TestComplex.hpp [hunk 2]
```diff
@@ -48,11 +48,6 @@
 
 namespace Test {
 
-#ifdef KOKKOS_COMPILER_NVHPC
-// warning: 'long double' is treated as 'double' in device code
-#pragma diag_suppress 20208
-#endif
```
Comment (dalg24) [inline core/unit_test/TestComplex.hpp:54]: This yielded
```
"/tmp/kokkos/core/unit_test/TestComplex.hpp", line 53: warning: invalid error number
  #pragma diag_suppress 20208
                        ^

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
Retest this please

File: .jenkins [hunk 3]
```diff
@@ -37,11 +37,20 @@ pipeline {
...
+                        OMP_PLACES = 'threads'
+                        OMP_PROC_BIND = 'spread'
+                        NVCC_WRAPPER_DEFAULT_COMPILER = 'nvc++'
+                    }
                     steps {
                         sh '''rm -rf build && mkdir -p build && cd build && \
                               /opt/cmake/bin/cmake \
                                 -DCMAKE_BUILD_TYPE=Debug \
-                                -DCMAKE_CXX_COMPILER=nvc++ \
+                                -DCMAKE_CXX_COMPILER=$WORKSPACE/bin/nvcc_wrapper \
```
Comment (masterleinad) [inline .jenkins:53]: Can you please clarify why this changed? I thought you wanted to test `nvc++` instead of `nvcc_wrapper` as `CUDA` compiler.

Comment (dalg24) [reply .jenkins:53]: With this PR, we would now be using nvc++ to compile CUDA code which is not quite fully ready.  So here I select the mvcc_wrapper with nvc++ as the host compiler to keep this build as it was before.

**Review by crtrott — APPROVED** [review]

File: .jenkins (hunk 3 above)
Comment (masterleinad) [reply .jenkins:53]: In this case, I would expect `KOKKOS_CXX_COMPILER_ID` to be `NVIDIA` and `KOKKOS_CXX_HOST_COMPILER_ID` to be `NVHPC` but all the changes here are for `KOKKOS_CXX_COMPILER_ID=NVHPC`.
What am I missing?

Comment (dalg24) [reply .jenkins:53]: Correct.  There is currently no build that test compiling with NVC++ but that is ok.
All the "do not rely on `ifdef __CUDA_ARCH__`" changes are getting us closer to it.  The main blocker is we need a major update to the Desul atomics.  Christian and I are looking into it.

Comment (masterleinad) [reply .jenkins:53]: I think in that case I would wait with this pull request until we can use `nvc++` as the device compiler in the CI. BUt I won't block if others want this to be merged already.

**Review by masterleinad — APPROVED** [review]
I understand that a number of people are actually using these changes to work on `nvc++` as device compiler support already.

FIne with me to move forward now (even if we don't really test the changes).

**Pull request merged**